iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,87 +17,83 @@
+        m_freeIndex = NO_INDEX;
+        return cxx::success<>();
+    }
+
+    // search from start
+    for (auto& entry : m_serviceDescriptions)
+    {
+        if (!entry)
         {
-            // Due to n:m communication we don't store twice but increase the reference counter
-            element.referenceCounter++;
+            entry.emplace(serviceDescription, 1U);
             return cxx::success<>();
         }
     }
-    uint64_t referenceCounter = 1U;
-    if (!m_serviceDescriptionVector.push_back({serviceDescription, referenceCounter}))
+
+    // append new entry at the end (the size only grows up to capacity)
+    if (m_serviceDescriptions.emplace_back())
     {
-        return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
+        auto& entry = m_serviceDescriptions.back();
+        entry.emplace(serviceDescription, 1U);
+        return cxx::success<>();
     }
 
-    auto serviceIndex = m_serviceDescriptionVector.size() - 1;
-    m_serviceMap.insert({serviceDescription.getServiceIDString(), serviceIndex});
-    m_instanceMap.insert({serviceDescription.getInstanceIDString(), serviceIndex});
-    m_eventMap.insert({serviceDescription.getEventIDString(), serviceIndex});
-    return cxx::success<>();
+    return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
 }
 
 void ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    bool removedElement{false};
-    uint64_t index = 0U;
-    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    auto index = find(serviceDescription);
+    if (index != NO_INDEX)
     {
-        auto& element = m_serviceDescriptionVector[index].serviceDescription;
-        if (element == serviceDescription)
+        auto& entry = m_serviceDescriptions[index];
+
+        if (entry->count > 1)
         {
-            auto& refCounter = m_serviceDescriptionVector[index].referenceCounter;
-            --refCounter;
-            if (refCounter == 0)
-            {
-                m_serviceDescriptionVector.erase(iterator);
-                removedElement = true;
-            }
-            else
-            {
-                return;
-            }
-            // There can't be more than one element
-            break;
+            entry->count--;
         }
-        ++index;
-        ++iterator;
-    }
-
-    auto removeIndexFromMap = [](std::multimap<capro::IdString_t, uint64_t>& map, uint64_t index) {
-        for (auto it = map.begin(); it != map.end();)
+        else
         {
-            if (it->second == index)
-            {
-                it = map.erase(it);
-                continue;
-            }
-            else if (it->second > index)
-            {
-                // update index due to removed element
-                it->second--;
-            }
-            it++;
+            entry.reset();
+            // reuse the slot in the next insertion
+            m_freeIndex = index;
```

Comment:
When you would use the fifo approach you could write here `m_freeIndices.push(index);`.

Reply:
Sure, but I will instead remove the optimization.